Application/ControlMsrE2/ControlMsrE2Hii.c

```diff
@@ -0,0 +1,397 @@
+                            (VOID *) &IfrVarStore->Guid,
+                            NULL,
+                            &DataSize,
+                            Data
+                            );
+
+            if (!EFI_ERROR (Status)) {
+              VarSize = sizeof (EFI_IFR_ONE_OF) - IfrOneOf->Header.Length;
+              VarSize = 8 - (VarSize / 3);
+
+              VarPointer = Data + IfrOneOf->Question.VarStoreInfo.VarOffset;
+              switch (VarSize) {
+              case 1:
+                VarStoreValue = *VarPointer;
+                break;
+              case 2:
+                VarStoreValue = *(UINT16 *) (VarPointer);
+                break;
+              case 4:
+                VarStoreValue = *(UINT32 *) (VarPointer);
+                break;
+              default:
+                VarStoreValue = *(UINT64 *) (VarPointer);
+                break;
+              }
+
+              Print (L" Value: value %X", VarStoreValue);
+            }
+            FreePool (Data);
+          }  ///< Allocate
+        }  ///< GetVariable
+        FreePool (VarStoreName);
+        Print (L"\n");
+      }
+    }
+  }
+  FreePool (HiiString);
+}
+
+VOID
+HandleIfrVariable (
+  IN OUT ONE_OF_CONTEXT *Context
+  )
+{
+  EFI_STATUS        Status;
+  EFI_STRING        HiiString;
+  UINTN             DataSize;
+  UINT8             *Data;
+  UINT32            Attributes;
+  UINTN             VarSize;
+  UINT8             *VarPointer;
+  UINT64            VarStoreValue;
+  UINT64            NewValue;
+
+  HiiString = HiiGetString (Context->EfiHandle, Context->IfrOneOf->Question.Header.Prompt, "en-US");
+  if (HiiString) {
+    Print (L"\nBIOS Option found: %s\n", HiiString);
+    FreePool (HiiString);
+  }
+
+  HiiString = AsciiStrCopyToUnicode ((CHAR8 *) Context->IfrVarStore->Name, 0);
+
+  if (HiiString == NULL) {
+    Print (L"\nCould not allocate memory for HiiString\n");
+    return;
+  }
+
+  VarSize = sizeof (EFI_IFR_ONE_OF) - Context->IfrOneOf->Header.Length;
+  VarSize = 8 - (VarSize / 3);
+
+  Print (
+    L"In VarStore \"%s\" GUID: %g Offset: %04X Size: %X ",
+    HiiString,
+    &Context->IfrVarStore->Guid,
+    Context->IfrOneOf->Question.VarStoreInfo.VarOffset,
+    VarSize
+    );
+
+  DataSize = 0;
+
+  Status = gRT->GetVariable (
+                  HiiString,
+                  (VOID *) &Context->IfrVarStore->Guid,
+                  &Attributes,
+                  &DataSize,
+                  NULL
+                  );
+
+  if (Status == EFI_BUFFER_TOO_SMALL) {
+    Data = AllocatePool (DataSize);
+    if (Data != NULL) {
+      Status = gRT->GetVariable (
+                      HiiString,
+                      (VOID *) &Context->IfrVarStore->Guid,
+                      &Attributes,
+                      &DataSize,
+                      Data
+                      );
+
+      if (!EFI_ERROR (Status)) {
+        VarPointer = Data + Context->IfrOneOf->Question.VarStoreInfo.VarOffset;
```

Comment:
Bounds are unchecked here!

Reply:
I want to replace the implementation here with https://github.com/acidanthera/audk/blob/master/MdeModulePkg/Library/VarCheckHiiLib/VarCheckHiiGen.c#L192, which should be more resilient. But I don't have resource to test it right now.

Reply:
But that does not check the bounds either? The point is, we do not trust data from NVRAM.